ComponentKit/HostingView/CKComponentHostingView.mm

```diff
@@ -130,33 +137,97 @@ - (void)componentScopeHandleWithIdentifier:(CKComponentScopeHandleIdentifier)glo
 {
   CKAssertMainThread();
   _pendingStateUpdates.insert({globalIdentifier, stateUpdate});
-  [self _setComponentNeedsUpdate];
+  [self _setNeedsUpdate:tryAsynchronousUpdate ? CKComponentHostingViewUpdateTypeAsynchronous : CKComponentHostingViewUpdateTypeSynchronous];
 }
 
 #pragma mark - Private
 
-- (void)_setComponentNeedsUpdate
+- (void)_setNeedsUpdate:(CKComponentHostingViewUpdateType)updateType
 {
-  if (!_componentNeedsUpdate) { // Avoid thrashing delegate
-    _componentNeedsUpdate = YES;
-    [self setNeedsLayout];
-    [_delegate componentHostingViewDidInvalidateSize:self];
+  if (updateType <= _componentNeedsUpdateType) {
+    return; // Never go backwards, e.g. from Sync to Async or Async to None.
+  }
+
+  _componentNeedsUpdateType = updateType;
+
+  switch (updateType) {
+    case CKComponentHostingViewUpdateTypeAsynchronous:
+      [self _asynchronouslyUpdateComponentIfNeeded];
+      break;
+    case CKComponentHostingViewUpdateTypeSynchronous:
+      [self setNeedsLayout];
+      [_delegate componentHostingViewDidInvalidateSize:self];
+      break;
+    case CKComponentHostingViewUpdateTypeNone:
+      break;
   }
 }
 
-- (void)_updateComponentIfNeeded
+- (void)_asynchronouslyUpdateComponentIfNeeded
+{
+  if (_scheduledAsynchronousComponentUpdate) {
+    return;
+  }
+  _scheduledAsynchronousComponentUpdate = YES;
+
+  // Wait until the end of the run loop so that if multiple async updates are triggered we don't thrash.
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [self _scheduleAsynchronousUpdate];
+  });
+}
+
+- (void)_scheduleAsynchronousUpdate
+{
+  if (_componentNeedsUpdateType != CKComponentHostingViewUpdateTypeAsynchronous) {
```

Comment:
I think the better way to check is `>`? If the update type is `none` then we're in bad territory and should assert.